### PR TITLE
DRAFT: Fix versioned package npm bug

### PR DESCRIPTION
**Pull request by shayneczyzewski**
# Description

This attempts to fix an `npm` bug which is present in the latest version of `node` (issue here: https://github.com/npm/cli/issues/5018).

Previously, we attempted to issue Prisma commands like so: `npx --no-install prisma@3.15.2 studio` This was because we wanted to ensure we did not use a globally installed one (possibly with a different version), nor ask users to install in the console if missing. But we also needed to be sure it would _somehow_ get installed on a clean project, so we had a fallback mechanism that if the command failed, it tried to `npm install` and then retried the failed command. All of this worked until that bug.

However, there was some progress in Wasp that makes some of this easier. We no longer need to worry about Prisma (or any NPM dep) not being installed, since we now `npm install` during compilation. Compilation happens before every command, so we can be sure we will have a local Prisma installation before we run any Prisma commands. Therefore, it feels pretty safe to use the local `node_modules` version. Thoughts? Risks?

## Type of change

Please select the option(s) that is more relevant.

- [ ] Code cleanup
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

**Comment by shayneczyzewski** [comment]
For some reason Mac CI is failing with: `Web app (stderr): wasp-cli: ./node_modules/.bin/prisma: streamingProcess: exec: invalid argument (Bad file descriptor)` when doing the migrate golden test, but they all run fine for me locally. 🤔

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -90,7 +90,7 @@ jobs:
       - name: Set up Node
         uses: actions/setup-node@v2
         with:
-          node-version: '16.15.0'
+          node-version: '16'
```
Comment (shayneczyzewski) [inline .github/workflows/ci.yaml:93]: Going back to latest LTS.

File: waspc/src/Wasp/Generator/Common.hs [hunk 2]
```diff
@@ -14,27 +14,16 @@ data ProjectRootDir
...
-    [ (SV.BackwardsCompatibleWith, latestLTSVersion),
-      (SV.LessThanOrEqual, latestLTSExactVersionThatWeKnowWorks)
-    ]
+  SV.rangeFromVersion (SV.BackwardsCompatibleWith, latestLTSVersion)
   where
-    latestLTSVersion = SV.Version 8 0 0 -- Goes with node 16
-    latestLTSExactVersionThatWeKnowWorks = SV.Version 8 5 5 -- Goes with node 16.15.0
+    latestLTSVersion = SV.Version 8 0 0 -- Goes with node 16 (but also higher versions too)
 
 prismaVersion :: SV.Version
```
Comment (shayneczyzewski) [inline waspc/src/Wasp/Generator/Common.hs:28]: Instead of a range, just using `BackwardsCompatibleWith` for both. However, for NPM it is slightly harder to pin down since 8.x is used across all versions of node.

File: waspc/src/Wasp/Generator/Common.hs [hunk 3]
```diff
@@ -14,27 +14,16 @@ data ProjectRootDir
 -- | Range of node versions that node packages generated by this generator work correctly with.
 nodeVersionRange :: SV.Range
 nodeVersionRange =
-  SV.rangeFromVersionsIntersection
-    [ (SV.BackwardsCompatibleWith, latestLTSVersion),
-      (SV.LessThanOrEqual, latestLTSExactVersionThatWeKnowWorks)
-    ]
+  SV.rangeFromVersion (SV.BackwardsCompatibleWith, latestLTSVersion)
```
Comment (sodic) [inline waspc/src/Wasp/Generator/Common.hs:17]: I wish Haskell had a consistent formatter. This can all fit in a single line, but it won't do it for you, but will allow it if you do it manually.

File: waspc/src/Wasp/Generator/DbGenerator/Jobs.hs [hunk 4]
```diff
@@ -42,115 +39,29 @@ migrateDev projectDir maybeMigrationName = do
   --   we are using `script` to trick Prisma into thinking it is running in TTY (interactively).
 
   -- NOTE(martin): For this to work on Mac, filepath in the list below must be as it is now - not wrapped in any quotes.
-  let npxPrismaMigrateCmd = npxPrismaCmd ++ ["migrate", "dev", "--schema", SP.toFilePath schemaFile] ++ optionalMigrationArgs
+  let npxPrismaMigrateCmd = absPrismaExecutableFp projectDir : ["migrate", "dev", "--schema", SP.toFilePath schemaFile] ++ optionalMigrationArgs
```
Comment (sodic) [inline waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:42]: We should change the variable's name to something like `prismaMigrateCmd` (since it no longer uses npx). The same goes for other `npxSomething` variable.

File: waspc/src/Wasp/Generator/Common.hs (hunk 3 above)
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/Common.hs:17]: Good catch! Yeah, that would be nice if it did try to tidy that up in some consistent way (but I also should have seen lol). Will update, thanks 👍🏻

File: waspc/src/Wasp/Generator/DbGenerator/Jobs.hs [hunk 5]
```diff
@@ -42,115 +39,29 @@ migrateDev projectDir maybeMigrationName = do
...
-data JobMessageForwardingStrategy = ForwardEverything | ForwardOnlyRetryErrors
-
-data Attempt = InitialAttempt | Retry
-
--- | Runs the original job, and if there is an error, it attempts recovery by running the second job
--- before retrying the original one last time.
--- TODO(shayne): Consider moving this function and data structures into a separate module to highlight it
--- is unique. We could also decouple this from DB-only jobs by passing in _jobType or using one of the
--- messages as a clue.
-retryJobOnErrorWith :: J.Job -> J.Job -> JobMessageForwardingStrategy -> J.Job
```
[the remark this answers is not in the document]
Comment (sodic) [reply waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:94]: F

File: web/docs/getting-started.md [hunk 6]
```diff
@@ -12,8 +12,8 @@ import useBaseUrl from '@docusaurus/useBaseUrl';
 ## 1. Requirements
 
 You need to have `node` (and `npm`) installed on your machine and available in `PATH`.
-- `node`: >=16.11.0 and <=16.15.0
-- `npm`: >= 8.0.0 and <= 8.5.5
+- `node`: >=16.11.0
+- `npm`: >= 8.0.0
```
Comment (sodic) [inline web/docs/getting-started.md:16]: From what I understood (and what the `ci.yml` file specifies), we no don't require node to be `>=16.11.0`. It's enough to have `16.x.x`.

The specifier `>=16.0.0` would also be incorrect since we don't allow `17.0.0` and higher. The best way to specify the version is either with `^16` or `16.x.x`. Perhaps it's easiest if we just took the version we had before [Martin's recent revision (old version on the left)](https://github.com/wasp-lang/wasp/commit/62f0a7693740aebdfdc094419bcbab4b86d7a488#diff-2abe1226319d07943a47ae2175fd39a81390105905ec26024c89d2a199d00711).

Btw, was the 11 minor version a mistake in the first place? Shouldn't it have been `>=16.0.0 and <=16.15.0`?

File: web/docs/getting-started.md [hunk 7]
```diff
@@ -33,12 +33,12 @@ We recommend using [nvm](https://github.com/nvm-sh/nvm) for managing your Node.j
 
   Then, install a version of node that you need, e.g.:
   ```shell-session
-  nvm install 16.15.0
+  nvm install 16.15.1
```
Comment (sodic) [inline web/docs/getting-started.md:36]: This should also probably be `nvm install 16`. NVM will install the newest stable version it finds by default, but if the user already has a lower compatible version (e.g., `16.5.0`), we don't want to force them to install `16.15.11` too. This part of the docs can also be taken directly from the older revision I linked in the previous comment.

The reason we had `16.15.0` was to ensure NVM wouldn't install `16.15.1` (although in hindsight we should have probably specified the correct range).

P.S. Even if we did want to insist on specifying the newest stable version explicitly, we'd have to update this document every couple of weeks :smile:

File: web/docs/getting-started.md [hunk 8]
```diff
@@ -33,12 +33,12 @@ We recommend using [nvm](https://github.com/nvm-sh/nvm) for managing your Node.j
...
   ```
 
   Finally, whenever you need to ensure specific version of node is used, run e.g.
   ```shell-session
-  nvm use 16.15.0
+  nvm use 16.15.1
```
Comment (sodic) [inline web/docs/getting-started.md:41]: Same story as `nvm install`, can also be taken from the older revision.

**Review by sodic — COMMENTED** [review]
You're pretty fast! I accidentally pushed a comment independent of the review, and you were done fixing it before I was done reviewing.

Anyway, nice thinking. I think the new strategy is perfectly legitimate. We just have to make sure to fix `getting-started.md` and it's ready for merging as far as I'm concerned.

File: waspc/src/Wasp/Generator/DbGenerator/Jobs.hs (hunk 4 above)
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:42]: Good catch, will update, thanks!

File: web/docs/getting-started.md (hunk 6 above)
Comment (shayneczyzewski) [reply web/docs/getting-started.md:16]: Arg, good point, I will update it to `16.x.x` for easier groking by people. Which was how it was specified before. I believe the reason it stated `16.11.0` was because that was the first version using `npm >= 8.0.0`.

File: web/docs/getting-started.md (hunk 7 above)
Comment (shayneczyzewski) [reply web/docs/getting-started.md:36]: Yeah good points all around :) I will use `nvm install 16` and `nvm use 16` like before 👍🏻

File: web/docs/getting-started.md (hunk 6 above)
Comment (shayneczyzewski) [reply web/docs/getting-started.md]: NOTE: I also updated npm to use `8.x.x`. Thanks

**Review by sodic — APPROVED** [review]
All good!

File: waspc/src/Wasp/Generator/DbGenerator/Jobs.hs [hunk 9]
```diff
@@ -42,115 +39,29 @@ migrateDev projectDir maybeMigrationName = do
...
 
-  retryJobOnErrorWith job (npmInstall projectDir) ForwardEverything
+  runNodeCommandAsJob serverDir "script" scriptArgs J.Db
 
 -- | Runs `prisma studio` - Prisma's db inspector.
 runStudio :: Path' Abs (Dir ProjectRootDir) -> J.Job
 runStudio projectDir = do
   let serverDir = projectDir </> serverRootDirInProjectRootDir
   let schemaFile = projectDir </> dbSchemaFileInProjectRootDir
+  let prismaStudioCmd = ["studio", "--schema", SP.toFilePath schemaFile]
```
Comment (Martinsos) [inline waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:57]:
```suggestion
  let prismaStudioCmdArgs = ["studio", "--schema", SP.toFilePath schemaFile]
```

File: waspc/src/Wasp/Generator/DbGenerator/Jobs.hs [hunk 10]
```diff
@@ -42,115 +39,29 @@ migrateDev projectDir maybeMigrationName = do
...
-  let job = runNodeCommandAsJob serverDir (head npxPrismaStudioCmd) (tail npxPrismaStudioCmd) J.Db
-
-  retryJobOnErrorWith job (npmInstall projectDir) ForwardEverything
+  runNodeCommandAsJob serverDir (absPrismaExecutableFp projectDir) prismaStudioCmd J.Db
 
 generatePrismaClient :: Path' Abs (Dir ProjectRootDir) -> J.Job
 generatePrismaClient projectDir = do
   let serverDir = projectDir </> serverRootDirInProjectRootDir
   let schemaFile = projectDir </> dbSchemaFileInProjectRootDir
+  let prismaGenerateCmd = ["generate", "--schema", SP.toFilePath schemaFile]
```
Comment (Martinsos) [inline waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:65]:
```suggestion
  let prismaGenerateCmdArgs = ["generate", "--schema", SP.toFilePath schemaFile]
```

File: waspc/src/Wasp/Generator/DbGenerator/Jobs.hs [hunk 11]
```diff
@@ -5,26 +5,23 @@ module Wasp.Generator.DbGenerator.Jobs
...
---    a user without prisma to install the latest version.
--- We also pin the version to what we need so it won't accidentally find a different version globally
---   somewhere on the PATH.
-npxPrismaCmd :: [String]
-npxPrismaCmd = ["npx", "--no-install", "prisma@" ++ show prismaVersion]
+-- | NOTE: The expectation is that `npm install` was already executed
+-- such that we can use the locally installed package.
+-- This assumption is ok since it happens during compilation now.
+prismaInServerNodeModules :: Path' (Rel ProjectRootDir) File'
+prismaInServerNodeModules = serverRootDirInProjectRootDir </> [SP.relfile|./node_modules/.bin/prisma|]
```
Comment (Martinsos) [inline waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:21]: Just to confirm -> so we don't use `npx` any more in order to avoid that bug, so instead we call the binary directly?
I guess it is tricky that this binary might move, right? with `npx prisma` we are sure we are going to get it, but here we are not I think? But I guess that is ok since we will catch it in tests anyway if it stops working?

**Review by Martinsos — APPROVED** [review]
Left 2 small comments, but looking good to me, and great that we are able to simplify this now! I wanted to mention that we need to reverse the changes I did to docs, but I see you already took care of that so that is it!

File: waspc/src/Wasp/Generator/DbGenerator/Jobs.hs (hunk 9 above)
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:57]: Good catch 👍🏻

File: waspc/src/Wasp/Generator/DbGenerator/Jobs.hs (hunk 11 above)
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:21]: Good question. I think the two main risks to this working now but breaking in the future are:
1) prisma executable changes name
2) prisma executable changes location

For the first, I think the fact that we pin a specific prisma version should prevent this. For the second, I think this would be a pretty disruptive change by npm for the ecosystem.

npx I am guessing just assumes (or uses) the `npm bin` output when resolving locally from within a project (like `.wasp/out/server`):
```bash
$ npx which prisma
/home/shayne/dev/wasp/waspc/examples/todoApp/.wasp/out/server/node_modules/.bin/prisma

$ npm bin
/home/shayne/dev/wasp/waspc/examples/todoApp/.wasp/out/server/node_modules/.bin
```

the `npm bin` output is the same on WSL, Windows, and Mac in that it just appends `node_modules/.bin` to PWD.

But yes, the TLDR is our e2e test should detect this if this ever changes as well. Worst case, if this does not work as well as we want, we could add a script that has more logic in it to find corner cases as they arise. That may re-invent `npx` but at least we'd allow people to use the current latest with this change.
